### PR TITLE
Quality of life improvements to payout processors

### PR DESCRIPTION
**Pull request by Kukks**
* Allows more fleixble intervals for payout processing from 10-60 mins to 1min-24hours(requested by users)
* Cancel ln payotus that expired (bolt11)
* Allow cancelling of ln payotus that have failed to be paid after x attempts
* Allow conifguring a threshold for when to process on-chain payouts (reduces fees)

### CODE REVIEW

**Review by NicolasDorier — APPROVED** [review]

**Review by NicolasDorier — CHANGES_REQUESTED** [review]
Need to protect `failedPayoutCounter` with lock.

**Comment by NicolasDorier** [comment]
I made a small commit which I think simplify the code. Feel free to revert.
Also I believe you need to lock the dictionary or use a ConcurrentDictionarry.

**Comment by NicolasDorier** [comment]
You need to add a test for `ProcessNewPayoutsInstantly`.

File: BTCPayServer.Tests/GreenfieldAPITests.cs [hunk 1]
```diff
@@ -3669,16 +3670,135 @@ await adminClient.GenerateOnChainWallet(admin.StoreId, "BTC", new GenerateOnChai
...
+            
+            //quick test: when updating processor, it processes instantly
+            await beforeHookTcs.Task.WaitAsync(TimeSpan.FromSeconds(5));
+            await afterHookTcs.Task.WaitAsync(TimeSpan.FromSeconds(5));
+            
+            settings =
+                Assert.Single(await adminClient.GetStoreOnChainAutomatedPayoutProcessors(admin.StoreId, "BTC"));
+            Assert.Equal(0.5m, settings.Threshold);
+            
+            //create a payout that should not be processed straight away
```
Comment (Kukks) [inline BTCPayServer.Tests/GreenfieldAPITests.cs:3751]:
```suggestion
            //create a payout that should not be processed straight away due to threshold
```

File: BTCPayServer/PayoutProcessors/BaseAutomatedPayoutProcessor.cs [hunk 2]
```diff
@@ -20,98 +21,136 @@ namespace BTCPayServer.PayoutProcessors;
...
         if (blob.Interval < TimeSpan.FromMinutes(AutomatedPayoutConstants.MinIntervalMinutes))
             blob.Interval = TimeSpan.FromMinutes(AutomatedPayoutConstants.MinIntervalMinutes);
         if (blob.Interval > TimeSpan.FromMinutes(AutomatedPayoutConstants.MaxIntervalMinutes))
             blob.Interval = TimeSpan.FromMinutes(AutomatedPayoutConstants.MaxIntervalMinutes);
-        await Task.Delay(blob.Interval, CancellationToken);
+        
+        _timerCTs??= new CancellationTokenSource();
+        try
+        {
+            await Task.Delay(blob.Interval, CancellationTokenSource.CreateLinkedTokenSource(CancellationToken, _timerCTs.Token).Token);
```
Comment (NicolasDorier) [inline BTCPayServer/PayoutProcessors/BaseAutomatedPayoutProcessor.cs:138]: I believe you need to dispose `CreateLinkedTokenSource`.

File: BTCPayServer/PayoutProcessors/BaseAutomatedPayoutProcessor.cs [hunk 3]
```diff
@@ -20,98 +21,136 @@ namespace BTCPayServer.PayoutProcessors;
...
+        }
+    }
+
+    private CancellationTokenSource _timerCTs;
+    private IEventAggregatorSubscription _subscription;
+
+    public void SkipInterval()
+    {
+        _timerCTs ??= new CancellationTokenSource();
+        _timerCTs?.Cancel();
```
Comment (NicolasDorier) [inline BTCPayServer/PayoutProcessors/BaseAutomatedPayoutProcessor.cs:151]: There is probably a concurrency issue here if two SkipInterval get called as same time

Comment (Kukks) [reply BTCPayServer/PayoutProcessors/BaseAutomatedPayoutProcessor.cs]: added a lock

File: BTCPayServer/PayoutProcessors/BaseAutomatedPayoutProcessor.cs (hunk 2 above)
Comment (Kukks) [inline BTCPayServer/PayoutProcessors/BaseAutomatedPayoutProcessor.cs:138]:
```suggestion
            var cts= CancellationTokenSource.CreateLinkedTokenSource(CancellationToken, _timerCTs.Token);
            await Task.Delay(blob.Interval, cts.Token);
            cts.Dispose();
```

**Pull request merged**